File: sdk/python/src/mewisme_meta/runtime.py
from __future__ import annotations

import asyncio
import contextlib
import json
import os
from dataclasses import dataclass
from pathlib import Path

from .distribution import RuntimeDistributionOptions, resolve_runtime_path
from .errors import ProtocolMismatchError, RuntimeLaunchError

PROTOCOL_MAJOR = 1
_MAX_BOOTSTRAP_BYTES = 64 << 10
_MAX_STDERR_BYTES = 16 << 10


@dataclass(frozen=True, slots=True)
class RuntimeBootstrap:
    endpoint: str
    token: str
    protocol: int


@dataclass(frozen=True, slots=True)
class ManagedRuntimeOptions(RuntimeDistributionOptions):
    listen: str = "127.0.0.1:0"
    token: str | None = None
    bootstrap_timeout: float = 10.0
    stop_timeout: float = 5.0


class ManagedRuntime:
    def __init__(
        self,
        process: asyncio.subprocess.Process,
        bootstrap: RuntimeBootstrap,
        stop_timeout: float,
        stderr_task: asyncio.Task[bytes],
    ) -> None:
        self._process = process
        self._stderr_task = stderr_task
        self.endpoint = bootstrap.endpoint
        self.token = bootstrap.token
        self.protocol = bootstrap.protocol
        self.stop_timeout = stop_timeout
        self._stopped = False

    @property
    def pid(self) -> int | None:
        return self._process.pid

    @classmethod
    async def start(cls, options: ManagedRuntimeOptions | None = None) -> ManagedRuntime:
        options = options or ManagedRuntimeOptions()
        env = dict(os.environ)
        if options.env:
            env.update(options.env)
        runtime_path = await asyncio.to_thread(resolve_runtime_path, options, env)
        if options.token:
            env["META_RUNTIME_TOKEN"] = options.token
        try:
            process = await asyncio.create_subprocess_exec(
                runtime_path,
                "--listen",
                options.listen,
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
                env=env,
            )
        except OSError as error:
            raise RuntimeLaunchError(f"failed to start {Path(runtime_path).name}") from error
        stderr_task = asyncio.create_task(_capture_stderr(process.stderr))
        try:
            bootstrap = await _read_bootstrap(process, options.bootstrap_timeout, stderr_task)
            if bootstrap.protocol != PROTOCOL_MAJOR:
                process.terminate()
                raise ProtocolMismatchError(PROTOCOL_MAJOR, bootstrap.protocol)
            return cls(process, bootstrap, options.stop_timeout, stderr_task)
        except BaseException:
            if process.returncode is None:
                process.terminate()
            with contextlib.suppress(Exception):
                await asyncio.wait_for(process.wait(), 1.0)
            stderr_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await stderr_task
            raise

    async def stop(self) -> None:
        if self._stopped:
            return
        self._stopped = True
        if self._process.returncode is None:
            self._process.terminate()
            try:
                await asyncio.wait_for(self._process.wait(), self.stop_timeout)
            except TimeoutError:
                self._process.kill()
                await self._process.wait()
        self._stderr_task.cancel()
        with contextlib.suppress(asyncio.CancelledError):
            await self._stderr_task


async def _capture_stderr(stream: asyncio.StreamReader | None) -> bytes:
    if stream is None:
        return b""
    data = bytearray()
    while True:
        chunk = await stream.read(4096)
        if not chunk:
            break
        data.extend(chunk)
        if len(data) > _MAX_STDERR_BYTES:
            del data[:-_MAX_STDERR_BYTES]
    return bytes(data)


async def _read_bootstrap(
    process: asyncio.subprocess.Process, timeout: float, stderr_task: asyncio.Task[bytes]
) -> RuntimeBootstrap:
    if process.stdout is None:
        raise RuntimeLaunchError("runtime stdout is unavailable")
    line_task = asyncio.create_task(process.stdout.readline())
    exit_task = asyncio.create_task(process.wait())
    try:
        done, _ = await asyncio.wait({line_task, exit_task}, timeout=timeout, return_when=asyncio.FIRST_COMPLETED)
        if line_task in done:
            try:
                line = line_task.result()
            except ValueError as error:
                raise RuntimeLaunchError("runtime bootstrap line is too large") from error
            if not line:
                details = await _stderr_text(stderr_task)
                raise RuntimeLaunchError(f"runtime exited before bootstrap{details}")
            if len(line) > _MAX_BOOTSTRAP_BYTES:
                raise RuntimeLaunchError("runtime bootstrap line is too large")
            try:
                value = json.loads(line)
            except (UnicodeDecodeError, json.JSONDecodeError) as error:
                raise RuntimeLaunchError("runtime emitted invalid bootstrap JSON") from error
            endpoint = value.get("endpoint")
            token = value.get("token")
            protocol = value.get("protocol")
            if (
                not isinstance(endpoint, str)
                or not endpoint
                or not isinstance(token, str)
                or not token
                or not isinstance(protocol, int)
            ):
                raise RuntimeLaunchError("runtime emitted invalid bootstrap fields")
            return RuntimeBootstrap(endpoint=endpoint, token=token, protocol=protocol)
        if exit_task in done:
            details = await _stderr_text(stderr_task)
            raise RuntimeLaunchError(f"runtime exited before bootstrap ({process.returncode}){details}")
        details = await _stderr_text(stderr_task) if stderr_task.done() else ""
        raise RuntimeLaunchError(f"runtime bootstrap timed out after {timeout:g}s{details}")
    finally:
        for task in (line_task, exit_task):
            if not task.done():
                task.cancel()


async def _stderr_text(task: asyncio.Task[bytes]) -> str:
    if not task.done():
        return ""
    data = await task
    text = data.decode("utf-8", errors="replace").strip()
    return f": {text}" if text else ""

File: sdk/node/src/runtime.ts
import {spawn, type ChildProcess} from "node:child_process"
import {basename} from "node:path"
import {setTimeout as delay} from "node:timers/promises"
import {ProtocolMismatchError, RuntimeLaunchError} from "./errors.js"

export const PROTOCOL_MAJOR = 1
const maxBootstrapBytes = 64 << 10
const maxStderrBytes = 16 << 10

export interface RuntimeBootstrap { endpoint: string; token: string; protocol: number }
export interface ManagedRuntimeOptions { runtimePath?: string; listen?: string; token?: string; env?: NodeJS.ProcessEnv; bootstrapTimeoutMs?: number; stopTimeoutMs?: number }

export class ManagedRuntime {
  readonly endpoint: string
  readonly token: string
  readonly protocol: number
  readonly pid: number | undefined
  private readonly child: ChildProcess
  private readonly stopTimeoutMs: number
  private stopped = false

  private constructor(child: ChildProcess, bootstrap: RuntimeBootstrap, stopTimeoutMs: number) {
    this.child = child
    this.endpoint = bootstrap.endpoint
    this.token = bootstrap.token
    this.protocol = bootstrap.protocol
    this.pid = child.pid
    this.stopTimeoutMs = stopTimeoutMs
  }

  static async start(options: ManagedRuntimeOptions = {}): Promise<ManagedRuntime> {
    const runtimePath = options.runtimePath ?? process.env.META_RUNTIME_PATH ?? (process.platform === "win32" ? "meta-runtime.exe" : "meta-runtime")
    const child = spawn(runtimePath, ["--listen", options.listen ?? "127.0.0.1:0"], {env: {...process.env, ...options.env, ...(options.token ? {META_RUNTIME_TOKEN: options.token} : {})}, stdio: ["ignore", "pipe", "pipe"], windowsHide: true})
    const stderr: Buffer[] = []
    let stderrBytes = 0
    child.stderr?.on("data", (chunk: Buffer) => {
      if (stderrBytes >= maxStderrBytes) return
      const kept = chunk.subarray(0, maxStderrBytes - stderrBytes)
      stderr.push(kept)
      stderrBytes += kept.length
    })
    try {
      const bootstrap = await readBootstrap(child, options.bootstrapTimeoutMs ?? 10_000, () => Buffer.concat(stderr).toString("utf8").trim())
      if (bootstrap.protocol !== PROTOCOL_MAJOR) {
        child.kill("SIGTERM")
        throw new ProtocolMismatchError(PROTOCOL_MAJOR, bootstrap.protocol)
      }
      return new ManagedRuntime(child, bootstrap, options.stopTimeoutMs ?? 5_000)
    } catch (error) {
      if (child.exitCode === null && child.signalCode === null) child.kill("SIGTERM")
      if (error instanceof ProtocolMismatchError || error instanceof RuntimeLaunchError) throw error
      throw new RuntimeLaunchError(`failed to start ${basename(runtimePath)}`, {cause: error})
    }
  }

  async stop(): Promise<void> {
    if (this.stopped) return
    this.stopped = true
    if (this.child.exitCode !== null || this.child.signalCode !== null) return
    const exited = new Promise<void>(resolve => this.child.once("exit", () => resolve()))
    this.child.kill("SIGTERM")
    if (await Promise.race([exited.then(() => true), delay(this.stopTimeoutMs).then(() => false)])) return
    this.child.kill("SIGKILL")
    await exited
  }
}

async function readBootstrap(child: ChildProcess, timeoutMs: number, stderr: () => string): Promise<RuntimeBootstrap> {
  const stdout = child.stdout
  if (!stdout) throw new RuntimeLaunchError("runtime stdout is unavailable")
  return new Promise<RuntimeBootstrap>((resolve, reject) => {
    let data = Buffer.alloc(0)
    let settled = false
    const finish = (error?: unknown, value?: RuntimeBootstrap) => {
      if (settled) return
      settled = true
      clearTimeout(timer)
      stdout.off("data", onData)
      child.off("error", onError)
      child.off("exit", onExit)
      if (error) reject(error)
      else resolve(value!)
    }
    const onData = (chunk: Buffer) => {
      data = Buffer.concat([data, chunk])
      if (data.length > maxBootstrapBytes) return finish(new RuntimeLaunchError("runtime bootstrap line is too large"))
      const newline = data.indexOf(0x0a)
      if (newline < 0) return
      const line = data.subarray(0, newline).toString("utf8").trim()
      try {
        const parsed = JSON.parse(line) as Partial<RuntimeBootstrap>
        if (!parsed.endpoint || !parsed.token || typeof parsed.protocol !== "number" || !Number.isInteger(parsed.protocol)) throw new Error("invalid bootstrap fields")
        finish(undefined, {endpoint: parsed.endpoint, token: parsed.token, protocol: parsed.protocol})
      } catch (error) {
        finish(new RuntimeLaunchError("runtime emitted invalid bootstrap JSON", {cause: error}))
      }
    }
    const onError = (error: Error) => finish(new RuntimeLaunchError(`runtime process failed: ${error.message}`, {cause: error}))
    const onExit = (code: number | null, signal: NodeJS.Signals | null) => finish(new RuntimeLaunchError(`runtime exited before bootstrap (${signal ?? code ?? "unknown"})${stderr() ? `: ${stderr()}` : ""}`))
    const timer = setTimeout(() => finish(new RuntimeLaunchError(`runtime bootstrap timed out after ${timeoutMs}ms${stderr() ? `: ${stderr()}` : ""}`)), timeoutMs)
    stdout.on("data", onData)
    child.once("error", onError)
    child.once("exit", onExit)
  })
}
